### DISCUSSION TITLE
Cleaning data from __NEXT_DATA__ for cached pages

### DISCUSSION THREAD

**Discussion by dtothefp** [vercel/next.js#29068] [Help]
@devknoll looks like the static method `renderDocument` was removed in your PR.

https://github.com/vercel/next.js/pull/28032

We are trying to update to Canary version of NextJS to fix the client routing issues and we were using `renderDocument` to clean some items from `__NEXT_DATA__` for pages we are caching at the CDN level. I'm sure this isn't probably best practice but do you advice of how to do something similar currently now that your changes have been merged in?

**Comment by devknoll** [comment]
Could you provide a more detailed example of what you were doing (when you were calling `renderDocument`) and what you were trying to clean and I could try to provide some guidance?

**Reply by dtothefp** [reply]
Sure, some of this is legacy code I'm trying to determine if there is still usefulness in cleaning up. From the props that were passed to renderDocument we were cleaning

- `props.__NEXT_DATA__.query`. at the CDN level we cache based upon an Accept List of query params.
- location data parsed from headers in` props.__NEXT_DATA__.props`. I don't think we really need to do this because we update the Vary header based upon location but like I said it is legacy code
- a user ID value within `props.__NEXT_DATA__.props`. this ID changes based upon user so any PII cached in `__NEXT_DATA__` won't be correct between server and client. the user id is generally persisted through a cookie.

Over all I think most of this cleaning of attributes is to reduce the possibility of error of using a cached value from the server render to hydrate client code. I'm not sure what if any best practices are outlined for such conditions in NextJS but if you have any recommendations it would be helpful.

The `renderDocument` call looked something like this 

```typescript
  static renderDocument(doc, props: DocumentProps): React.ReactElement {
    const cleanProps = cleanNextData(props as RenderDocumentProps);
    return Document.renderDocument<DocumentPageProps>(doc, cleanProps);
  }
```

**Reply by devknoll** [reply]
Thank you for the context. What I would generally recommend doing instead is changing your call sites (e.g. `getInitialProps` or `getServerSideProps`, whichever you are using) to either clean the props before returning, or setting the appropriate `Cache-Control` header to prevent your CDN from caching. If your `pages/_app` has `getInitialProps` too, make sure it gets updated too.

Let me know if you don't think that will work for you.

**Reply by dtothefp** [reply]
Thanks for the info. Yes we are using `getServerSideProps` for multiple pages and `getInitialProps` for a custom `_document`. WE are cleaning parameters there but what was being cleaned before was `__NEXT_DATA__.query`. From my understanding after your change there would be no way to change that value via NextJS source code on our end. Like I said it's probably something we shouldn't be doing but just wanted to check

<img width="983" alt="Screen Shot 2021-09-14 at 2 26 14 PM" src="https://user-images.githubusercontent.com/4656726/133465514-542461a6-a85c-475d-acc1-3b55b4814472.png">

**Reply by devknoll** [reply]
Yeah, query parameters are trickier. I’m not sure it makes sense to handle them within Next.js: CDNs should have a way for you to configure how query strings are handled, including either not forwarding them, or forwarding them but using them as part of the cache key.

The most that I would recommend doing within Next.js is checking if the request has any query parameters that you don’t want cached and if so, setting `Cache-Control` appropriately. Mutating `__NEXT_DATA__` isn’t really safe or recommended, you should really clean up the request before it gets to Next (via CDN, another reverse proxy, or a custom server).

Hope that helps.
